### ISSUE TITLE
Fastclear Bug With Intel Mesa Adapters on the GL Backend

### ISSUE THREAD

**Issue by zicklag** [gfx-rs/wgpu#1627]
**Description**
When using the OpenGL backend on Linux the clear color seems to behaving strangely. For one, the color itself is lighter than it should be. For two, the borders around any objects drawn on top of the clear color have an even lighter, pixelated version of the clear color.

I'm currently troubleshooting but I opened the issue to start the discussion. With my experimentation so far it seems completely related to the act of clearing the draw framebuffer. All other rendering and examples seem to work fine, and if you have an example where you can't see the clear color, such as the skybox example, everything looks great.

Also I've noticed some weird related behavior in my Renderdoc captures:

When I launch the example with Renderdoc it has the same problem as running it without Renderdoc ( which makes sense ):

![image](https://user-images.githubusercontent.com/25393315/125171019-93aedb00-e177-11eb-96ff-d92d8f33f6c9.png)

And when I view the renderbuffer contents after the initial clear in renderdoc, it shows the clear color like it shows in the render, lighter than it should be:

![image](https://user-images.githubusercontent.com/25393315/125171112-1041b980-e178-11eb-8ddd-801e4ea2ed29.png)

The clear color shows the same in the draw step, all the way until the final framebuffer blit, where it is dark enough ( but still with pixelated edges ):

![image](https://user-images.githubusercontent.com/25393315/125171162-572faf00-e178-11eb-9ec7-68af3fd79c2d.png)

Yet, when I hover over the pixels in the image, the little thumbnail at the bottom shows the wrong ligher color:

![image](https://user-images.githubusercontent.com/25393315/125171260-d8874180-e178-11eb-9697-59d496201f61.png)

Very strange.

I found that I could get rid of the pixelated edges by forcing the renderbuffer pixel format to be `RGBA8`, but the color was still off. I think thats the closest lead I have and I'm going to look into how different pixel formats efffect it, and maybe try binding the framebuffer storage to a texture instead of a renderbuffer and see if that makes any difference.

---

**PS:** Very excited that the new GL backend on `wgpu-hal` is working for all the examples! This is the first time I've tried it that the `shadow`, `boids`,  and `skybox` examples have worked. I might try to tackle #1617, but I figured I'd try to get this one out of the way first. :)

**Repro steps**
Run the `cube` or `shadow` example with the OpenGL backend.

**Expected vs observed behavior**
There should be no pixelated edges around objects and the clear color should be darker.

**Expected:**
![image](https://user-images.githubusercontent.com/25393315/125170894-e4720400-e176-11eb-9bfc-55e1b36fc259.png)

**Actual ( I think it's fine if the lines for the trangles aren't there, the issue is the background color ) :**
![image](https://user-images.githubusercontent.com/25393315/125170723-ea1b1a00-e175-11eb-93a6-faf0bec15b5f.png)

**Extra materials**

**Shadow example:**
![image](https://user-images.githubusercontent.com/25393315/125170729-f43d1880-e175-11eb-815f-cbfd290b37de.png)

**Platform**
Running WGPU `cube` or `shadow` examples on Linux Pop!_OS ( Ubuntu ) 20.04. Adapter info:

```
[2021-07-10T16:54:11Z INFO  wgpu_hal::gles::adapter] Vendor: Intel
[2021-07-10T16:54:11Z INFO  wgpu_hal::gles::adapter] Renderer: Mesa Intel(R) UHD Graphics (CML GT2)
[2021-07-10T16:54:11Z INFO  wgpu_hal::gles::adapter] Version: OpenGL ES 3.2 Mesa 20.0.8
[2021-07-10T16:54:11Z INFO  wgpu_hal::gles::adapter] SL version: OpenGL ES GLSL ES 3.20
```

**Comment by zicklag** [comment]
I'm just realizing that the lighter color of the clear color is actually because the entire render is actually lighter than it should be, not just the clear color.

By forcing the pixel format of the renderbuffer to `RGB8`, that fixes the pixelated edges. Then when I do a renderdoc capture and I got to the last step of the frame, I can actually save the texture for the final image and it looks exactly how we want it, but when we run the actual example in the window, it looks wrong. I wonder if the color problem is a compositing thing, and the pixelated edges is a pixel format thing.

**Export From Renderdoc:**

![test](https://user-images.githubusercontent.com/25393315/125172172-bfcd5a80-e17d-11eb-84f2-7a5e1961e318.png)


**Screenshot of app ( for the same capture as above renderdoc export ):**

![image](https://user-images.githubusercontent.com/25393315/125172204-f7d49d80-e17d-11eb-8859-4ec4c7eb218d.png)

**Comment by zicklag** [comment]
Hey, I think I might have gotten it! #1628. Not 100% sure it's a final solution, but we can continue discussion in the PR.

![image](https://user-images.githubusercontent.com/25393315/125173960-c660cf80-e187-11eb-8c9c-11db725976df.png)

**Comment by cwfitzgerald** [comment]
So this is actually a long standing bug on Intel cards on Linux with srgb. I'm not sure how we actually should be working around this reasonably, but it likely needs to be internal with shader rewriting. I'll take a look at your pr later tonight.

**Comment by kvark** [comment]
@cwfitzgerald do you have a reference to the bug somewhere? Is this related to `COLORSPACE_SRGB`, or to the alpha channel writes that are blocked by #1628 ?

**Comment by cwfitzgerald** [comment]
@kvark I've never actually filed this as I first hit it back when I was a baby graphics programmer and didn't know how to do such things :) I need to collect the information around it (as I suspect #725 is related) and file it up to the appropriate places. Maybe I'll hit up mesa on IRC now that that's a thing I know I can do.

@zicklag Thank you for filing this! It is a very detailed issue which is always much appreciated!

So a couple housekeeping things first. Do the reftests work correctly on gl on your machine? You can test this by running `WGPU_BACKEND=gl cargo test --example <example>`. If they do, this indicates it is something related to swapchain shenanigans. 

So looking at the images, it looks like what you put as the "expected" image is actually completely missing gamma correction and the erroneous image properly has error correction (though with artifacts). I can reproduce this lack of srgb conversion on my intel/linux machine on both vulkan and GL. I also confirmed on a separate machine that the darker image shows that there is no proper SRGB transformation going on when there is supposed to be.

So this is 100% a driver bug at this point. As a user, you can so srgb conversion in your fragment shader or tonemapping pass with a regular framebuffer and it will work as expected.

There won't be a simple fix for this unfortunately. Basically we're going need to:
- Detect we're on intel/mesa.
- Lie about creating an srgb swapchain.
- Inject srgb conversion into all shaders that are writing to the swapchain.

This has some major hurtles that needs to be crossed first:
- We need to figure out standard ways for us to implement driver bug workarounds.
- Naga needs to be able to inject srgb conversion.
- We need to have the ability to have more than one backend shader program per wgpu pipeline. This is because, at pipeline creation, we don't know if the pipeline is going to be used to rendering to a swapchain or not. We need to decide at pipeline bind time.

**Comment by kvark** [comment]
Injecting shader code isn't very practical for this, since we wouldn't be able to blend in proper space, anyway.
So let's reach out to Intel/Mesa and see if there is a better workaround. In the worst case, we can disable presentation entirely (i.e. say that no adapter is compatible to a surface) for this platform.

**Comment by cwfitzgerald** [comment]
So the blocking around the triangle is caused by https://gitlab.freedesktop.org/mesa/mesa/-/issues/2565. Can you run with the env flag `INTEL_DEBUG=nofc` and see if the bug goes away? This isn't a long term solution, but shows that it is the problem.

The issue about vk not properly doing srgb seems unrelated.

**Comment by cwfitzgerald** [comment]
@zicklag are you using wayland?

**Comment by cwfitzgerald** [comment]
Interesting, the original issue was different behavior on wayland.

**Comment by zicklag** [comment]
> Thanks for testing!

No problem! :D

> Do the reftests work correctly on gl on your machine?

No. Neither the `shadow` or `cube` examples will pass the reftest for OpenGL.

> So looking at the images, it looks like what you put as the "expected" image is actually completely missing gamma correction and the erroneous image properly has error correction (though with artifacts).

Oh, that's interesting. Makes sense now that I look at the screenshots in the example dirs. :) The reftests for Vulkan will interestingly still pass despite the sRGB difference. Not sure if that's expected.

> Can you run with the env flag INTEL_DEBUG=nofc and see if the bug goes away?

That fixed it! Nice to know that something so simple was actually a driver bug and not something wrong with the code. I didn't write the code, but I couldn't figure out for the life of me why a `gl.clear_buffer()` would be so weird. Also, the reftests will pass with that environment variable set as well.

![image](https://user-images.githubusercontent.com/25393315/125344278-b2020b80-e31c-11eb-97d6-43b6cbf047e6.png)

---

So, in summary, there are 3 separate issues here, if I understand correctly:

- **Vulkan not doing sRGB conversion on Linux**: This is happening for me on my Linux machine
- **GL not doing sRGB conversions on Linux**: @cwfitzgerald did you say you were getting the darker image even on GL on your machine? I'm not getting that on my machine for GL. Just for Vulkan.
- **Blocky clear borders on GL:** This is because of a mesa bug, and can short term be worked around by setting an environment variable.

That leaves the actionable items to be:

- Figure out how to ensure sRGB conversion is done for Vulkan and GL on Linux
- Figure out how we want to workaround the mesa bug on GL

Does that sound about right?

**Comment by zicklag** [comment]
Another data point I just realized is that in my experiment to port Bevy's latest rendering branch from WGPU 0.9 to WGPU master I found out that everything looked darker, which is probably because of the gamma correction/sRGB conversion problem ( I don't know the difference :grin:  )

I could possibly try to do a git bisect to see what commit caused it. Though it might be some massive commit like "switch to wgpu-hal" or something...

**WGPU 0.9:**

![image](https://user-images.githubusercontent.com/25393315/125347169-386c1c80-e320-11eb-944b-22a7210ca38c.png)

**WGPU master:**

![image](https://user-images.githubusercontent.com/25393315/125347436-897c1080-e320-11eb-9dc8-521117678440.png)

**Comment by zicklag** [comment]
Yeah, the last time that I can find vulkan gamma working correctly was this commit: 19206064dc7779546e3d9ade9f8093935da5613e. And the first time I can find it broken ( where it also compiles on Linux ) is this commit: 5578222685d0f949e60e48bbcec46179aeab7bbf.

It was the move to using `wgpu-hal` for vulkan support, so no small, easy update to find. I'll have to try to compare `gfx`'s vulkan backend and how `wgpu-hal` differs in the instance creation or something.

**Comment by cwfitzgerald** [comment]
> No. Neither the shadow or cube examples will pass the reftest for OpenGL.

Could you upload what the files `wgpu/examples/<name>/screenshot-actual.png` and `screenshot-difference.png` are for the failing tests? This will show why they are failing. Both _should_ pass on the current GL impl.

> The reftests for Vulkan will interestingly still pass despite the sRGB difference. Not sure if that's expected.

Yeah I think this is a swapchain thing, the reftests are headless so don't deal with the swapchain.

> GL not doing sRGB conversions on Linux

Yeah my haswell rig has srgb issues on both vk and gl.

> Does that sound about right?

Yup! Thank you for this amazingly clear issue!

> It was the move to using wgpu-hal for vulkan support

Yeah I kinda feared that. Thanks for taking a look at the difference. I'm surprised this is only manifesting on intel though.

**Comment by zicklag** [comment]
> Could you upload what the files wgpu/examples/<name>/screenshot-actual.png and screenshot-difference.png are for the failing tests? This will show why they are failing. Both should pass on the current GL impl.

Yep. It's just failing because of the intel fastclear bug. When passing `INTEL_DEBUG=nofc` it works fine.

( [screenshot-actual](https://user-images.githubusercontent.com/25393315/125357545-6015b180-e32d-11eb-81fa-c59afc3136c7.png), [screenshot-difference](https://user-images.githubusercontent.com/25393315/125357546-60ae4800-e32d-11eb-8b3c-58ef51d504cf.png) )

> Yeah my haswell rig has srgb issues on both vk and gl.

Interesting. Could it have to do with the EGL version, maybe? It seems that sRGB is only enabled for egl > 1.5:

https://github.com/gfx-rs/wgpu/blob/c41efb02560deab02038d59a637451fdc9e525c1/wgpu-hal/src/gles/egl.rs#L541-L545

**Comment by cwfitzgerald** [comment]
oh. yeah that would do it XD

This is gonna be fun to try to work around, though GL already relies on a blit, so it should be possible.

**Comment by cwfitzgerald** [comment]
So I can reproduce the srgb issue on intel/vulkan/windows, so issue 1 is definitely a vulkan backend issue.

**Comment by zicklag** [comment]
> So I can reproduce the srgb issue on intel/vulkan/windows, so issue 1 is definitely a vulkan backend issue.

Yeah, I'm still looking around for what might cause it, but I haven't found anything yet. The swapchain creation looks nearly exactly the same as the gfx backend, so I'm looking around other places but nothing has stood out to me.

---

Let me know if you want to open separate issues for these, I don't care either way, but on the topic of the Intel fastclear bug, from my tests we can avoid it by setting the environment variable inside the GL backend at instance creation, but unfortunately the environment variable needs to be set before we load the X11 ( and I'm assuming wayland, too ) display. This means we can't do it while enumerating adapters, which is where it would be nicest to do it because we could wait to add the environment variable until we found out that they were using an Intel Mesa adapter.

Do you think it's safe enough just to set the `INTEL_DEBUG=nofc` variable at instance creation for now and do that regardless of what device you are using?

**Comment by cwfitzgerald** [comment]
Yeh we probably should split out the bugs into separate issues, this one could be for the fast clear issue. If you can do it, that'd be great, otherwise I'll do it a bit later today.

> Do you think it's safe enough just to set the INTEL_DEBUG=nofc variable at instance creation for now and do that regardless of what device you are using?

Yeah this should be fine, as long as we always set it on linux. It shouldn't affect any other adapters as it's an env only for the intel cards. This is a fine (if slow) short term solution, but we should work to see if we can find a way to prevent the bug from occurring at all because this is pessimising intel cards w/o the bug.

**Comment by zicklag** [comment]
Great. I opened #1645 with the short-term solution, and I renamed this issue to be specific to the fastclear bug. I'll open a new issue for the Vulkan sRGB bug.

**Comment by cwfitzgerald** [comment]
So I think we should do the following for dealing with the fastclear bug:

1. We should counter this by doing an explicit shader clear when do you a clear on affected hardware (run a fullscreen triangle that just outputs the clear color). This is what a slow clear is actually doing (and why fastclears are so much faster) so should be no worse.
2. We can tell if WebGL is affected by using the `WEBGL_debug_renderer_info` extension to get [`UNMASKED_RENDERER_WEBGL` and `UNMASKED_VENDOR_WEBGL`](https://developer.mozilla.org/en-US/docs/Web/API/WebGL_API/Constants#constants_defined_in_webgl_extensions). We should be using this to get the device/vendor information anyway. You can check this on your own hardware though https://webglreport.com/?v=2.

**Comment by zicklag** [comment]
Just finished taking both of those measures and it now succesfully works around the bug on both desktop and WebGL. :tada: Currently it's in my WebGL branch. I'm not sure if it's helpful or not, but let me know if you want me to split it out to a separate PR for just the desktop fastclear fix instead of leaving it merged with WebGL.

**Comment by cwfitzgerald** [comment]
Amazing! Yes splitting it out would be preferred, it'll make it easier to review.

**Comment by zicklag** [comment]
There you go! Created #1717. It turned out a little messier than I had hoped because, in order to draw the triangle for the shader clear, I had to add a bunch of boolean state values to keep track of whether `gl::DEPTH_TEST` and friends were currently enabled or disabled so that I could re-set those values back to whatever they were after disabling them all to draw the triangle.

I'm not sure if there's a simpler way to do that, but it's all I could come up with. Let me know if you have any ideas!
